server/src/utils/auth.utils.ts

```diff
@@ -0,0 +1,92 @@
+import { Types } from 'mongoose';
+import bcrypt from 'bcryptjs';
+import crypto from 'crypto';
+import * as jwt from 'jsonwebtoken';
+import { OAuth2Client, LoginTicket, TokenPayload } from 'google-auth-library';
+import { JWT_MAX_AGE, JWT_MIN_AGE } from '../constants';
+import { User } from '../models/user.model';
+import { UserAuthSchema } from '../types/user.types';
+
+const hashInput = async function (input: string): Promise<string> {
+  const salt = await bcrypt.genSalt(10);
+  return await bcrypt.hash(input, salt);
+};
+
+const validateInput = function (input: string, hashedInput: string): Promise<boolean> {
+  return bcrypt.compare(input, hashedInput);
+};
+
+const generateRandomToken = (length: number): string => {
+  return crypto.randomBytes(length || 32).toString('hex');
+};
+
+const generateJwt = (
+  userId: Types.ObjectId,
+  email: string,
+  accessType: string,
+  isRememberPassword: boolean
+): string => {
+  try {
+    return jwt.sign(
+      {
+        id: userId,
+        emailAddress: email,
+        accessType: accessType,
+      },
+      process.env.JWT_KEY as jwt.Secret,
+      { expiresIn: isRememberPassword ? JWT_MAX_AGE : JWT_MIN_AGE }
+    );
+  } catch (error: any) {
+    throw new Error('Invalid token or no authentication present');
+  }
+};
+
+const verifyJwt = async (token: string): Promise<UserAuthSchema> => {
+  if (token == null || token === '') {
+    throw new Error('Unauthorized');
+  }
+
+  try {
+    const valid: jwt.JwtPayload = jwt.verify(
+      token,
+      process.env.JWT_KEY as jwt.Secret
+    ) as jwt.JwtPayload;
+    if (valid == null) throw new Error('Forbidden');
+
+    // at this point should return a user object
+    // with id, email and accessType
+    const userObj: UserAuthSchema | null = await User.findOne({ _id: valid.id })
+      .select('-createdAt -updatedAt -password')
```

Comment:
LOVE THIS!!! save data

Reply:
Haha I think I can actually remove more fields, but gotta check if we only need userId, email, and accessType. if so I'll remove all other fields too